server/gameconnection.py

```diff
@@ -294,36 +296,54 @@ async def handle_game_result(self, army, result):
         except (KeyError, ValueError):  # pragma: no cover
             self._logger.warning("Invalid result for %s reported: %s", army, result)
 
-    async def handle_operation_complete(self, army, secondary, delta):
-        if not int(army) == 1:
+    async def handle_operation_complete(
+        self, primary: Any, secondary: Any, delta: str
```

Comment:
I think I said these would be strings before, but after looking at the code again, I realized they can be any json value. (Have a look at `handle_action` and `LobbyConnection.on_message_received`). The GPGNet messages look like this:

```json
{
    "command": "OperationComplete",
    "target": "game",
    "args": [1, 1, "12:34"]
}
```
The parameters to these `handle_*` functions are just the args array from the message, so the type can be anything produced by `json.loads`. Which is also the reason why you'll see lots of calls to `str` and `int` on those arguments in handler functions.

Reply:
As someone coming from the world of django-rest-framework it might be nice to pass these messages through some kind of standard serializer so that the data is always ready for the database rather than each handler having to do this?

Reply:
I'm not entirely sure what you mean. The messages are already deserialized by the json parser. But the message format is essentially a remote function call, you can see it just sends an arg list. The most standard way to automatically convert types would probably be to deserialize to named tuples instead and define the types that way, but then you have to define a named tuple for every handler function which seems bad.

I suppose we could write some decorator or something that inspects the type arguments of the handler function and calls appropriate parsing functions. That would work pretty well for primitives I think, in which case you could just call the parameter type like: `parameter.type(arg)` e.g. `str(arg)` but if the params are tuples or something, then it would be much more annoying logic to write. You'd have to handle each case separately I think.

Seems like a lot of work, but it would be pretty cool.